### DISCUSSION TITLE
Install node-red or other packages on side of openmptcprouter

### DISCUSSION THREAD

**Discussion by doryphoreanonyme** [Ysurac/openmptcprouter#2255] [Q&A]
Hello Everyone,

I am planning to use opentmptcp router to transmit video from a remote location with a couplke 4G modems.

This part works great and I'm really happy with openmptcprouter. 

However, in order to take profit of the VPN I would like to use the raspberry pi's GPIO for controlling video equipment.

I'm unfortunately unable to install components such Node-Red with curl or even python Rpi-GPIO, because apt is not installed on the image. 

Does Anyone can help me understand how I could add the neccessary packages to install such componenents?

Thanks a lot for your help!

**Comment by Ysurac** [comment]
OpenMPTCProuter is based on OpenWrt, to install packet opkg is used.
I don't know if there is Node-Red or Rpi-GPIO packaged for OpenWrt.

**Reply by doryphoreanonyme** [reply]
Thanks for your so quick answer, a quick research with OpenWrt I found that GPIO are accesible with bash scrtipting: https://openwrt.org/docs/techref/hardware/port.gpio

Which could already be a helper for me. Because I'm also trying to have a heartbeat LED on the system on one GPIO to properly handle power supply and shutdown. Worst case I could use a second pi for control.

However the dtoverlay doesnt seems to work for LED, but is ok for shutdown from a pin. Could it be related?

There is what i added to config.txt and is working fine on rasbian lite:
# led heartbeat
dtoverlay=gpio-led,gpio=23,trigger=heartbeat
# pin shutdown
dtoverlay=gpio-shutdown,gpio_pin=24,gpio_pull=off,active_low=1,debounce=200

again thanks for your great work!

**Reply by doryphoreanonyme** [reply]
For thoses how might land here:

With Ysursac answer, I was able to search for nodes packages for openWrt. So after a lot of trial and errors I was able to find how it works.
1. Update your packages list with 
opkg update 

2. You can list the packages availables (optional)
Opkg list

3. You can then install node.js and npm (node package manager)
Opkg install node npm

finally install node-red (cmd from their website)
npm install -g --unsafe-perm node-red

5. Run Node-Red as explain on their website

At this point GPIOs doesn't seems accesible by node red. But YAY, node-red is running!

**Comment by Adorfer** [comment]
i strongly discourage to overload the OMR with additional tasks. 
You basically 

1) add complexity to your setup 
2) make it harder to update the openmptcp-stack (risk of breaking other stuff)
3) run into the risk of regressions if e.g. future OMRrelated kernel patches affect the performance of your side installs. 

considering nodered: 
it tends to be mass storage intense. don't do that on the same MLC flash your OMR resides on (eg. raspi on microsd or usb-key) unless you aim for scheduled flashcard swap ever n months.
